src/mongocxx/test/spec/unified_tests/operations.cpp

```diff
@@ -553,6 +553,112 @@ document::value fail_point(entity::map& map, spec::apm_checker& apm, document::v
                          kvp("failPoint", args["failPoint"]["configureFailPoint"].get_string()));
 }
 
+document::value find_one_and_delete(collection& coll,
+                                    client_session* session,
+                                    document::view operation) {
+    document::view arguments = operation["arguments"].get_document().value;
+    document::view filter = arguments["filter"].get_document().value;
+    options::find_one_and_delete options{};
+
+    if (arguments["collation"]) {
+        options.collation(arguments["collation"].get_document().value);
+    }
+    if (arguments["hint"]) {
+        if (arguments["hint"].type() == bsoncxx::v_noabi::type::k_string)
+            options.hint(hint{arguments["hint"].get_string().value});
+        else
+            options.hint(hint{arguments["hint"].get_document().value});
+    }
+    if (arguments["projection"]) {
+        options.projection(arguments["projection"].get_document().value);
+    }
+    if (arguments["sort"]) {
+        options.sort(arguments["sort"].get_document().value);
+    }
+
+    stdx::optional<document::value> document;
+    if (session) {
+        document = coll.find_one_and_delete(*session, filter, options);
+    } else {
+        document = coll.find_one_and_delete(filter, options);
+    }
+
+    // Server versions below 3.0 sometimes return an empty document rather than null when no
```

Comment:
Thanks so much for looking into this! Your explanation and repro make sense to me, but is this a bug in the C++ or C driver? Should I link to some ticket for a `TODO` to remove the logic around expecting an empty document instead of `null`? Happy to file a ticket.

Reply:
I think the C driver should document that the reply is always initialized to the empty document for unacknowledged writes, (as it does for bulk writes). Created https://github.com/mongodb/mongo-c-driver/pull/789.

I think it is arguable whether the C++ driver should be changed to return a `std::nullopt` instead of an empty document for an unacknowledged write. And I think I'm in favor of not changing it.
- Right now, there is no ambiguity. `std::nullopt` unambiguously means that nothing was matched. For an unacknowledged write, there may or may not have been a match, but the empty document unambiguously signals that there is no server response.
- Changing it is a slight backwards break in the return.

Reply:
Can you update this comment, and the others, to note that this only applies to unacknowledged writes (and not the server version)?
```suggestion
    // Using an unacknowledged write returns an empty document rather than null when no
```

Reply:
Gotcha. The documentation for the C driver looks good to me, and I agree that changing the CXX driver to always return `std::nullopt` is backwards-breaking and not really necessary. Updated comments.